packages/dds/migration-shim/src/sharedTreeShimFactory.ts

```diff
@@ -0,0 +1,79 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+import {
+	type IChannelAttributes,
+	type IFluidDataStoreRuntime,
+	type IChannelServices,
+	type IChannelFactory,
+} from "@fluidframework/datastore-definitions";
+
+import { type SharedTreeFactory } from "@fluid-experimental/tree2";
+import { SharedTreeShim } from "./sharedTreeShim";
+import { attributesMatch } from "./utils";
+
+/**
+ * {@link @fluidframework/datastore-definitions#IChannelFactory} for {@link SharedTreeShim}.
+ *
+ * Creates the migration shim that allows a migration from legacy shared tree to shared tree.
+ * Note: There may be a need for 3 different factories for different parts of the migration.
+ * That or three different shims. Potentially we can just do 2 as 2 and 3 can be combined.
+ * 1. pre-migration
+ * 2. after a summary has been generated but there may still be potential v1 ops
+ * 3. post-migration after a summary has been generated and the msn has moved far enough forward for only v2 ops
+ *
+ * @sealed
+ */
+export class SharedTreeShimFactory implements IChannelFactory {
```

Comment:
Having two separate factories (as opposed to e.g. a single bimodal factory) implies that the data object or container author is registering both, I think (so that the appropriate one can be used based on... attributes I think)?  Because there is no way to know the migration state of the document pre-load.

I think it may be easier to go for a multi-modal shim factory (which can e.g. inspect blobs from the snapshot to determine the correct configuration of the runtime object pre-instantiation) than  to rely on factory switching (and changing attributes post-creation, which historically hasn't been done).

Reply:
As an example, here's merge-tree reading a blob from the snapshot to determine whether to proceed loading a legacy vs. V1 snapshot, we could use a similar distinction to determine whether to create a legacy ST or new ST:

https://github.com/microsoft/FluidFramework/blob/main/packages/dds/merge-tree/src/snapshotLoader.ts#L47

Reply:
The function that is shared amongst all `IChannelContext` classes is `loadChannelFactoryAndAttributes`

Reply:
I don't think you need to change anything at that layer - e.g. merge-tree is able to do this bimodal load from within a single factory without any of those changes to channel context.

Reply:
We can implement a more generic solution with two factories as since we change the type we can allow the migration from `SharedCell` to `SharedMap` or `SharedMap` to `SharedDirectory`. If we stay on the same type then we can't reuse most of this code to migrate from `SharedMap` to `SharedDirectory`. We would need to either ship some code that makes the Shim very complicated, or create a new object.

Reply:
Furthermore, the `IChannelContext` is the one that summarizes the `attributes` blob. So we would need to put it on a different blob. And we would want to pick a specific `fluidMetadata` blob as I have no idea what names `SharedTree` and `LegacySharedTree` use as blob names.

Reply:
We chatted about this and will go with separate factories/shims/etc. to get the initial package stood up, and we can always reevaluate later if we learn more.